[Dynamo] Preserve source_fn in node.meta

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92399


Sample value from the test case `test_export_with_stack_trace`

node.target | node.meta["source_fn"]
-- | --
aten.randn.default | <built-in method randn of type object at 0x7f8683263108>
aten.t.default | < built-in function linear >
aten.mm.default | < built-in function linear >
aten.cos.default | <built-in method cos of type object at 0x7f8683263108>
aten.relu.default | relu
aten.add.Tensor | < built-in function add >


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

SherlockNoMad committed:
Preserve source torch api calls in node.meta

[ghstack-poisoned]

jerryzh168 commented on test/dynamo/test_export.py:
```diff
@@ -894,6 +895,7 @@ def forward(self, x):
             if node.op not in {"placeholder", "output"}:
                 self.assertTrue(node.stack_trace is not None)
                 self.assertTrue(node.meta["nn_module_stack"] is not None)
+                self.assertTrue(node.meta["torch_api_call"] is not None)
```
should we test what it is as well? same for "nn_module_stack"

jerryzh168 reviewed (approved):
thanks a lot!

SherlockNoMad commented on torch/_dynamo/output_graph.py:
```diff
@@ -778,6 +778,10 @@ def create_proxy(
 
+        torch_api_call = tx.torch_api_call
+        if torch_api_call:
+            rv.node.meta["torch_api_call"] = torch_api_call
```
Need a better name for this. I don't want this only to be used for torch.* function. I suspect we will need to extend this for other buildin.functions and tensor.methods(). 

alternative names: original_fn, source_fn

jerryzh168 replied:
makes sense, we want to have things like `operator.add` here as well

SherlockNoMad replied:
Update to source_fn.

jerryzh168 replied:
btw is `operator.add` supported now?

SherlockNoMad replied:
I simplified the implementation. We can directly store the original node.target in node.meta["source_fn"], and propagate it along lowering. 

See updated summary. 

This now also cover the case for operator.add.

SherlockNoMad committed:
Update on "[Dynamo] Preserve source torch api calls in node.meta"



Sample value from the test case `test_export_with_stack_trace`

node.target | node.meta["torch_api_call"]
-- | --
aten.randn.default | torch.randn
aten.t.default | torch._C._nn.linear
aten.mm.default | torch._C._nn.linear
aten.cos.default | torch.cos



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot rebase -s

SherlockNoMad committed:
Update on "[Dynamo] Preserve source_fn in node.meta"



Sample value from the test case `test_export_with_stack_trace`

node.target | node.meta["source_fn"]
-- | --
aten.randn.default | torch.randn
aten.t.default | torch._C._nn.linear
aten.mm.default | torch._C._nn.linear
aten.cos.default | torch.cos



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad committed:
Update on "[Dynamo] Preserve source_fn in node.meta"



Sample value from the test case `test_export_with_stack_trace`

node.target | node.meta["source_fn"]
-- | --
aten.randn.default | torch.randn
aten.t.default | torch._C._nn.linear
aten.mm.default | torch._C._nn.linear
aten.cos.default | torch.cos



cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

SherlockNoMad commented:
@pytorchbot merge -g
